graph-viz/app/algorithms/ForceBundling.java

```diff
@@ -0,0 +1,409 @@
+package algorithms;
+
+import models.EdgeVector;
+import models.Path;
+import models.Point;
+
+import java.util.ArrayList;
+
+/**
+ * Force Directed Edge Bundling Algorithm.
+ */
+public class ForceBundling {
+    /** TODO replace EdgeVector to be Edge, and
+     * use the point from edge class to get the position
+     * from point data structure.
+     */
+    // All the data nodes
+    private ArrayList<Point> dataNodes;
+    // All the data edges
+    private ArrayList<EdgeVector> dataEdges;
+    // Turning points of the generated paths
+    private ArrayList<Path> subdivisionPoints = new ArrayList<>();
+    // Compatible edge indexes for each edge
+    private ArrayList<ArrayList<Integer>> compatibilityList = new ArrayList<>();
+    // Algorithm parameters
+    // K: global bundling constant controlling edge stiffness
+    private final double K = 0.1;
+    // S: init. distance to move points
+    private double S_initial = 0.02;
+    // P_initial: init. subdivision number
+    private final int P_initial = 1;
+    // P_rate: subdivision rate increase
+    private final int P_rate = 2;
+    // C: number of cycles to perform
+    private final double C = 3;
+    // I_initial: init. number of iterations for cycle
+    private final double I_initial = 50;
+    // I_rate: rate at which iteration number decreases i.e. 2/3
+    private final double I_rate = 2.0 / 3.0;
+    // compatibility_threshold: the threshold score of deciding compatibility
+    private final double compatibility_threshold = 0.6;
+    // epsilon: decide the precision
+    private final double eps = 1e-6;
+    // isolatedEdgesCnt: unbundled edge count
+    private int isolatedEdgesCnt = 0;
+
+    /**
+     * Constructor of fdeb algorithm.
+     * @param dataNodes incoming data nodes.
+     * @param dataEdges incoming data edges.
+     */
+    public ForceBundling(ArrayList<Point> dataNodes, ArrayList<EdgeVector> dataEdges) {
+        this.dataNodes = dataNodes;
+        this.dataEdges = dataEdges;
+    }
+
+    public int getIsolatedEdgesCnt() {
+        return isolatedEdgesCnt;
+    }
+
+    /**
+     * Sets different moving distance to each zoom level.
+     * @param zoom current zoom level.
+     */
+    public void setS(int zoom) {
+        S_initial = Math.max(0.025 - zoom * 0.0025, 0.001);
+    }
+
+    private double vectorDotProduct(Point p, Point q) {
+        return p.getX() * q.getX() + p.getY() * q.getY();
+    }
+
+    /**
+     * Calculates the vector format of edge.
+     * @param e current edge.
+     * @return corresponding vector.
+     */
+    private Point edgeAsVector(EdgeVector e) {
+        return new Point(dataNodes.get(e.getTargetNodeInd()).getX() - dataNodes.get(e.getSourceNodeInd()).getX(), dataNodes.get(e.getTargetNodeInd()).getY() - dataNodes.get(e.getSourceNodeInd()).getY());
```

Comment:
Once you use Point Edge class instead of an EdgeVector and you would also have a point attribute in the Edge class, then you would not need